### PR TITLE
[cuDNN SDPA] combine mask with bias to support public SDPA API

### PR DESCRIPTION
**Pull request by Cjkkkk**
* cuDNN SDPA does not support mask input any more, therefore we combine the bias and mask manually to align with public SDPA API design.

### CODE REVIEW

**Comment by Cjkkkk** [comment]
@superbobry @kaixih

File: tests/fused_attention_stablehlo_test.py [hunk 1]
```diff
@@ -313,6 +308,7 @@ def test_sdpa_inference(self):
 
   @jtu.run_on_devices("cuda")
   def test_sdpa_var_seq(self):
+    self.skipTest("Skip before fixed.")
```
Comment (superbobry) [inline tests/fused_attention_stablehlo_test.py:311]: Is there a bug we can reference here? If not, can you add more context on why the test is skipped to the message string, please?

File: jax/_src/cudnn/fused_attention_stablehlo.py [hunk 2]
```diff
@@ -73,6 +73,15 @@ def should_export_dbias(bias_shape, query_shape, layout) -> bool:
     _, _, q_N, _ = query_shape
   return b_B == 1 and b_N == q_N
 
+def get_large_negative_number(dtype):
+  # temp WAR as cuDNN has a bug for subtraction between two large negative value
```
Comment (superbobry) [inline jax/_src/cudnn/fused_attention_stablehlo.py:77]: When can this workaround be removed? Is there a bug you can link to here?

File: jax/_src/cudnn/fused_attention_stablehlo.py [hunk 3]
```diff
@@ -73,6 +73,15 @@ def should_export_dbias(bias_shape, query_shape, layout) -> bool:
...
+  if dtype == jnp.bfloat16:
+    return jnp.asarray(-2 << 40, dtype=dtype)
```
Comment (superbobry) [inline jax/_src/cudnn/fused_attention_stablehlo.py:79]: Can we use -.7 * finfo.maxn as the previous version of this function did?

File: jax/_src/cudnn/fused_attention_stablehlo.py [hunk 4]
```diff
@@ -73,6 +73,15 @@ def should_export_dbias(bias_shape, query_shape, layout) -> bool:
...
+  elif dtype == jnp.float16:
+    return jnp.asarray(-2 << 14, dtype=dtype)
+  else:
+    raise ValueError("Unsupported dtype for inputs.")
```
Comment (superbobry) [inline jax/_src/cudnn/fused_attention_stablehlo.py:83]: It looks like the previous version of this function accepted more dtypes. Why is it okay to only allow b?float16 here?

Comment (Cjkkkk) [reply jax/_src/cudnn/fused_attention_stablehlo.py:83]: cuDNN Flash attention only support these 2 data type. Fp8 flash attention is not integrated yet.

File: jax/_src/cudnn/fused_attention_stablehlo.py (hunk 2 above)
Comment (Cjkkkk) [reply jax/_src/cudnn/fused_attention_stablehlo.py:77]: It is still not fixed in cuDNN AFAIK. So when cuDNN subtract two large negative number (softmax for example) will result in nan. There is an internal nvbug but nothing I can link here.

File: jax/_src/cudnn/fused_attention_stablehlo.py (hunk 3 above)
Comment (Cjkkkk) [reply jax/_src/cudnn/fused_attention_stablehlo.py:79]: I tried -.7 * finfo.maxn and it has the nan issue and cuDNN team recommended the -2 << 40 as WAR for now.

File: tests/fused_attention_stablehlo_test.py (hunk 1 above)
Comment (Cjkkkk) [reply tests/fused_attention_stablehlo_test.py:311]: I don't have more info now as I just noticed that with cuDNN 9.2 this test fail with less than 1% value diff larger than 1e05.

**Comment by superbobry** [comment]
I think the minimum cuDNN version JAX supports is 9.0. Does it make sense to remove mask entirely from the JAX API, given that?

CC @hawkinsp

**Comment by Cjkkkk** [comment]
> I think the minimum cuDNN version JAX supports is 9.0. Does it make sense to remove mask entirely from the JAX API, given that?
> 
> CC @hawkinsp

Removing mask completely should be fine as long as it aligned with the public API.

**Review by superbobry — APPROVED** [review]

**Comment by superbobry** [comment]
Please fix the type checker errors.

**Comment by superbobry** [comment]
Thanks, you just need to squash the commits and the PR is good to go.

**Comment by Cjkkkk** [comment]
hi, @superbobry is the pr good to go?
